### PR TITLE
GNN Boundary Layer: local mesh message-passing for surface pressure accuracy

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The Transolver slice attention is a **global operator** — it pools all mesh nodes into learned slices, computes slice-to-slice interactions, and broadcasts back. This is powerful for capturing global flow patterns but structurally wrong for **boundary layer physics**: near-wall pressure/velocity gradients are governed by local PDE propagation along the wall, at length scales orders of magnitude smaller than the global domain.

A small Graph Neural Network (GNN) module operating **only on surface and near-surface nodes** implements local message-passing that respects the boundary layer's inherent locality. After the Transolver backbone produces hidden states for all nodes, 2 rounds of GraphSAGE-style message passing among (a) the surface nodes and (b) their k=4 nearest volume neighbors lets near-wall information propagate locally **before** the SRF head applies its correction.

**This is fundamentally different from all previous approaches**: every prior inter-foil coupling attempt (fore-aft cross-attention, GALE, cross-DSDF) used global attention. The GNN operates locally on the wall, like a PDE solver taking small propagation steps.

**Literature support:** B-GNNs (arXiv:2503.18638, 2025) demonstrated 85% error reduction using GNN message passing with local physics-informed inputs on airfoil meshes. GraphSAGE aggregation (Hamilton et al., NeurIPS 2017) is the right choice for irregular CFD meshes.

**Target metrics:** p_tan (aft-foil tandem suction peak accuracy), p_in (leading-edge pressure)

---

### Step 1: Add config flags

In the `Config` class, add:
```python
gnn_boundary_layer: bool = False   # enable GNN message-passing on surface/near-wall nodes
gnn_layers: int = 2                # number of GNN message-passing rounds
gnn_k_neighbors: int = 4          # k nearest volume neighbors per surface node
```

### Step 2: Implement the BoundaryLayerGNN module

```python
class BoundaryLayerGNN(nn.Module):
    def __init__(self, n_hidden: int = 192, n_layers: int = 2):
        super().__init__()
        self.n_layers = n_layers
        self.layers = nn.ModuleList([
            nn.Sequential(
                nn.Linear(n_hidden * 2, n_hidden),
                nn.SiLU(),
                nn.Linear(n_hidden, n_hidden),
            )
            for _ in range(n_layers)
        ])
        self.layer_norms = nn.ModuleList([nn.LayerNorm(n_hidden) for _ in range(n_layers)])
        # Zero-init output of last layer for safe residual start
        nn.init.zeros_(self.layers[-1][-1].weight)
        nn.init.zeros_(self.layers[-1][-1].bias)

    def forward(self, hidden, surf_idx, neighbor_idx):
        """
        hidden: [B, N, C]
        surf_idx: [M_s] surface node indices
        neighbor_idx: [B, M_s, k] k nearest volume neighbors per surface node
        """
        h = hidden
        for layer, ln in zip(self.layers, self.layer_norms):
            h_surf = h[:, surf_idx, :]          # [B, M_s, C]
            h_nbr = h[:, neighbor_idx, :]        # [B, M_s, k, C]  (advanced indexing)
            h_agg = h_nbr.mean(dim=2)            # [B, M_s, C]
            h_in = torch.cat([h_surf, h_agg], dim=-1)  # [B, M_s, 2C]
            delta = layer(h_in)                  # [B, M_s, C]
            h_surf_new = ln(h_surf + delta)
            h = h.clone()
            h[:, surf_idx, :] = h_surf_new
        return h
```

### Step 3: Precompute neighbor indices

```python
def compute_surface_neighbors(x_coords, surf_idx, k=4):
    """
    x_coords: [B, N, 2]  node xy coordinates
    surf_idx: [M_s]
    Returns neighbor_idx: [B, M_s, k] — k nearest volume node indices per surface node
    """
    B, N, _ = x_coords.shape
    all_idx = torch.arange(N, device=x_coords.device)
    vol_mask = torch.ones(N, dtype=torch.bool, device=x_coords.device)
    vol_mask[surf_idx] = False
    vol_idx = all_idx[vol_mask]  # [M_v]

    surf_coords = x_coords[:, surf_idx, :2]   # [B, M_s, 2]
    vol_coords  = x_coords[:, vol_idx,  :2]   # [B, M_v, 2]

    # If M_v is too large (>2000), subsample 512 random volume nodes for efficiency
    if vol_idx.shape[0] > 2000:
        perm = torch.randperm(vol_idx.shape[0], device=x_coords.device)[:512]
        vol_idx = vol_idx[perm]
        vol_coords = x_coords[:, vol_idx, :2]

    diff = surf_coords.unsqueeze(2) - vol_coords.unsqueeze(1)  # [B, M_s, M_v, 2]
    dists = diff.pow(2).sum(-1)                                  # [B, M_s, M_v]
    k_actual = min(k, vol_idx.shape[0])
    _, knn_local = dists.topk(k_actual, dim=-1, largest=False)  # [B, M_s, k]
    return vol_idx[knn_local]  # [B, M_s, k]
```

**IMPORTANT efficiency note:** Call this once per batch (not per step). The mesh topology is fixed within a batch. Cache by reusing across steps if batch composition doesn't change.

### Step 4: Integrate into Transolver forward pass

In `Transolver.__init__`, add:
```python
if cfg.gnn_boundary_layer:
    self.gnn_bl = BoundaryLayerGNN(n_hidden=cfg.n_hidden, n_layers=cfg.gnn_layers)
```

In `Transolver.forward`, after the main backbone loop and **before** the SRF head:
```python
if self.cfg.gnn_boundary_layer and surf_idx is not None:
    neighbor_idx = compute_surface_neighbors(x[:, :, :2], surf_idx, k=self.cfg.gnn_k_neighbors)
    fx = self.gnn_bl(fx, surf_idx, neighbor_idx)
```

### Step 5: Add argparse flags

```python
parser.add_argument('--gnn_boundary_layer', action='store_true')
parser.add_argument('--gnn_layers', type=int, default=2)
parser.add_argument('--gnn_k_neighbors', type=int, default=4)
```

### Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent frieren \
  --wandb_name "frieren/gnn-boundary-layer-s42" \
  --wandb_group "frieren/gnn-boundary-layer" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --gnn_boundary_layer --gnn_layers 2 --gnn_k_neighbors 4

# Seed 73
cd cfd_tandemfoil && python train.py --agent frieren \
  --wandb_name "frieren/gnn-boundary-layer-s73" \
  --wandb_group "frieren/gnn-boundary-layer" \
  --seed 73 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --gnn_boundary_layer --gnn_layers 2 --gnn_k_neighbors 4
```

### Report

Table: p_in, p_oodc, p_tan, p_re for both seeds and 2-seed avg vs baseline. W&B run IDs. Note VRAM usage.

**Watch for:** If neighbor computation is slow, cap M_v to 512 random volume nodes (already in the code above). If VRAM spikes, use surface-only message passing (k nearest surface neighbors only, simpler).

---

## Baseline (PR #2251, 2-seed avg)

| Metric | Value | Target to beat |
|--------|-------|----------------|
| p_in | 11.891 | < 11.89 |
| p_oodc | 7.561 | < 7.56 |
| p_tan | 28.118 | < 28.12 |
| p_re | 6.364 | < 6.36 |

Baseline W&B runs: `7jix2jkg` (s42), `epkfhxfl` (s73)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### Metrics (Surface MAE pressure, lower is better)

| Metric | Baseline (2-seed avg) | GNN-BL s42 | GNN-BL s73 | GNN-BL 2-seed avg | Δ vs baseline |
|--------|----------------------|-------------|-------------|-------------------|---------------|
| p_in   | 11.891               | 13.760      | 14.808      | 14.284            | +2.393 (+20.1%) |
| p_oodc | 7.561                | 9.407       | 9.344       | 9.375             | +1.814 (+24.0%) |
| p_tan  | 28.118               | 28.619      | 31.487      | 30.053            | +1.935 (+6.9%)  |
| p_re   | 6.364                | 7.642       | 8.082       | 7.862             | +1.498 (+23.5%) |

### Run details

- **W&B run IDs**: `6du84780` (s42), `izedn36s` (s73)
- **W&B group**: `frieren/gnn-boundary-layer`
- **Epochs**: 121 (s42), 122 (s73) — hit 180-min timeout
- **Epoch time**: ~89s (vs ~78s baseline — ~14% slower due to per-batch k-NN computation)
- **Peak VRAM**: ~34.5 GB (measured from debug run)
- **GNN params**: 222,720 (2 GraphSAGE layers, k=4 neighbors)

### Command used

```bash
cd cfd_tandemfoil && python train.py --agent frieren \
  --wandb_name "frieren/gnn-boundary-layer-s42" \
  --wandb_group "frieren/gnn-boundary-layer" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --gnn_boundary_layer --gnn_layers 2 --gnn_k_neighbors 4
```

### What happened

**Negative result.** The GNN boundary layer degraded performance across all 4 validation splits, with the worst impacts on p_in (+20.1%), p_oodc (+24.0%), and p_re (+23.5%). Even p_tan — the target metric for inter-foil coupling — worsened by 6.9%.

### Analysis

1. **Hidden feature disruption**: The GNN operates on the backbone's hidden representations between the non-last blocks and the last block. Even with zero-init on the last GNN layer, the GNN's learned updates after training shift the feature distribution that the last TransolverBlock and SRF heads were jointly optimized to process. The GNN essentially inserts an untrained bottleneck in a well-tuned pipeline.

2. **Redundancy with existing SRF heads**: The Surface Refinement Head and Aft-Foil SRF head already perform local surface correction using hidden features. The GNN adds another layer of surface-specific processing that competes with rather than complements the SRF heads. The existing SRF approach (direct MLP correction per surface node) is simpler and more targeted than message-passing.

3. **k-NN overhead without proportional benefit**: The per-batch k-NN computation adds ~14% epoch time overhead. The volume-to-surface message passing aggregates features from k=4 nearest volume neighbors, but these neighbors contribute information that the Transolver's global slice attention already captures. The locality inductive bias doesn't help because the Transolver backbone is already good at extracting local features through spatial bias and gap-stagger conditioning.

4. **Training efficiency**: The 222K extra parameters (on top of ~2.7M base) are not large, but they add optimization pressure. The GNN parameters are in the "other_params" group with the full learning rate, which may be too aggressive for a module that should start near-identity.

### Suggested follow-ups

- **GNN on SRF hidden features only**: Instead of modifying backbone hidden states, apply GNN message passing *within* the SRF head — enrich surface node features with volume context before the correction MLP. This avoids disrupting the backbone-to-last-block pathway.
- **Frozen GNN warm-start**: Freeze GNN parameters for the first 50+ epochs, only unfreezing after the rest of the model is well-trained. This prevents early GNN updates from shifting the feature distribution.
- **Reduce GNN learning rate**: Use a 10x lower LR for GNN parameters (separate param group) to slow adaptation and keep outputs near-identity longer.
- **Surface-only message passing** (no volume neighbors): Pass messages between neighboring surface nodes only, which better matches the B-GNN literature finding that boundary layer accuracy is primarily about surface-local information exchange.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2269 because ADVISOR: Closing — all 4 metrics regressed significantly (p_in +20.1%, p_oodc +24.0%, p_tan +6.9%, p_re +23.5%).

**Root cause:** The GNN inserts an untrained bottleneck between the Transolver backbone and the SRF heads, disrupting a well-tuned feature distribution. Even with zero-init on the last GNN layer, trained updates shift what the last TransolverBlock and SRF heads expect to see. The GNN also competes with the existing SRF heads which already perform local surface correction — it's redundant.

**Key takeaway:** The Transolver's global slice attention + SRF pipeline is already capturing local boundary layer physics effectively. Adding an explicit GNN for local propagation adds optimization pressure without new information. The k-NN volume neighbors contribute context that slice attention already captures. Added to DO NOT REVISIT.

Good analysis and suggested follow-ups noted — but the magnitude of regression (+20-24%) indicates the approach is fundamentally incompatible with the existing pipeline, not just a hyperparameter issue. Moving you to a fresh bold direction..
